Help: Add track pings to contact form buttons

jeremeylduvall opened the pull request:
This PR adds a Tracks event to the "How can we help?" and "Mind sharing how you feel?" selectors on [the Contact page](http://wordpress.com/help/contact). The specific event being added is `calypso_help_contact_form_click`, and we're also tracking the value of the button in a prop called `button_value`. 

To test: 

1. Load http://wordpress.com/help/contact
2. Click one of the selectors underneath either field

The event is recorded (I use `localStorage.setItem('debug', 'calypso:analytics');` in the console to confirm) with the appropriate prop.

Of note, it won't fire if users just leave the default selections on the form, which is fine in my mind. This way, we can see how many users are using the form alongside what they're selecting.

Fix for #1822.

dllh commented:
The code looks good and I confirm that events are recording, but I wonder if the naming could be improved. The name `calypso_help_contact_form_click` seems over-general to me. For example, a click on the "Chat with Us" button is also a click on the contact form but isn't covered by this event. Would `calypso_help_contact_form_selector_click` be more descriptive (I guess that'd leave out clicks on the site selector, though)? Or should we make a distinct event name for each of the two selectors, named appropriately? So: `calypso_help_how_can_we_help_click` and `calypso_help_how_you_feel_click`, though maybe those are <em>too</em> specific and verbose.

jeremeylduvall commented:
I went back and forth on the event naming. I ultimately thought we could make use of the props for separating out the events instead of creating different event names for each set of buttons. Separating out the events would probably make life much easier moving forward though. Will work on that today.

dllh commented:
I waffled on it a bit too. One piece of information I think we lose with the current name, and which is part of the basis for my suggesting a modified approach, is which _set_ of buttons was clicked. Let's say we decide to start providing different options in the two selectors. And then let's further say that we decide to test the efficacy of various sorts of selectors, or of selector order. Or say we want to see if a category selector vs. an emotion selector gets more clicks. To track meaningfully across these various other possible dimensions, I think we need either a new name or an extra property to denote the extra info. So if you want a generic name still, but would like to capture more context, you could add a property named `selector_label` and values `How can we help` and `Mind sharing how you feel` (or, in the future, possibly something like `Feature Category`) and then also store the value as at present (though I'd advocate for a name more like `selector_value` or `selected_option`). This'd let us query the two sets of values separately without having to rely on continuity of the values over time (e.g., what if we later added some other field that had a value overlapping with one of the ones we're using here? we would muddy our understanding of what was actually clicked where, or we'd have to track what changed over time, vs. having specificity built into the event properties).

jeremeylduvall commented:
@dllh give it a go now. I moved everything into a function `trackClickStats` that fires on click. Depending on the `selectionName`, the events are now called `calypso_help_how_can_we_help_click` or `calypso_help_how_you_feel_click`. Both events record the option value in `selected_option`. Ultimately, I think this will set us up better down the road.

dllh commented:
This looks reasonable to me, but I'd rather get other eyes on it too to make sure we're not breaking any standards or other best practices, which I'm less familiar with than many developers. @jordwest mind giving this a look?

jordwest commented on client/me/help/help-contact-form/index.jsx:
```diff
@@ -94,6 +95,18 @@ module.exports = React.createClass( {
+			} );
+		}
+	},
+
```
Just a style thing, but this could also be written with a switch statement:

```js
switch ( selectionName ) {
    case 'howCanWeHelp':
        return analytics.tracks.recordEvent( 'calypso_help_how_can_we_help_click', {
            selected_option: option.value
        } );
    case 'howYouFeel':
        return analytics.tracks.recordEvent( 'calypso_help_how_you_feel_click', {
            selected_option: option.value
        } );
}
```

Or, easier to extend if we add more options later:

```js
const tracksEvent = {
    howCanWeHelp: 'calypso_help_how_can_we_help_click',
    howYouFeel: 'calypso_help_how_you_feel_click'
}[ selectionName ];

if ( tracksEvent ) {
    analytics.tracks.recordEvent( tracksEvent, { selected_option: option.value } );
}
```

jordwest commented on client/me/help/help-contact-form/index.jsx:
```diff
@@ -94,6 +95,18 @@ module.exports = React.createClass( {
 		this.setState( { siteSlug } );
 	},
 
+	trackClickStats: function( selectionName, option ) {
```
Here we could be more specific - since the `trackClickStats` function doesn't need the whole `option` object, we could just accept the `selected_option` directly:

```js
trackClickStats: function( selectionName, select_option ) {
   // ...
}
```

jordwest commented:
I've left a bit of feedback on the diffs (I edited them some so please read them on GitHub if you're reading the notification emails).

@dllh raised some concerns about using variables in the event tracking function call as I've suggested above, rather than plain strings. It looks like it's been done that way in [other parts of the codebase](https://github.com/Automattic/wp-calypso/blob/d2a47fd38981f8da4d95d52c3126ec8e52ea5c35/client/lib/plugins/actions.js#L81) though so we should be safe (blame me if it breaks something :smile:). I couldn't find any reference to an automated code scan done on track events so I'm not sure we're restricted in the way we are with i18n strings.

dllh commented:
Now that I think about it some more, I think the guideline re event names is that they be spelled out literally in the code rather than pieced together from parts (e.g. `'foo_' + someVarName + '_bar'), so what's used here should be fine. I do like the less repetitive code that results from using variables.

jeremeylduvall commented:
Thanks for the feedback @jordwest! Switched the if block to use the second (const) method you mentioned. Squashed everything into one commit. Still works as expected. Want to take a look?

dllh commented:
I like this. Will give the last word to @jordwest, but consider it a :+1: from me.

jordwest commented:
Thanks for making those changes, LGTM :+1:

jordwest commented on client/me/help/help-contact-form/index.jsx:
```diff
@@ -116,7 +128,8 @@ module.exports = React.createClass( {
 				onClick: () => {
-					this.setState( { [ selectionName ]: option.value } )
+					this.setState( { [ selectionName ]: option.value } ),
+					this.trackClickStats( selectionName, option.value )
```
Just one more minor thing, each line should be ended with a semicolon as this is a function body rather than an object:
this.setState( { [ selectionName ]: option.value } );
this.trackClickStats( selectionName, option.value );

jeremeylduvall commented:
Sorry — slipped past me :) @jordwest mind giving this one last sanity check?

jordwest commented:
:ship:
